episodes/instance-segmentation-and-measurements.md

```diff
@@ -0,0 +1,596 @@
+function.
+
+```python
+
+from skimage.segmentation import clear_border
+
+viewer.layers.remove('instance_seg')
+
+instance_seg = clear_border(instance_seg)
+
+viewer.add_labels(instance_seg)
+```
+
+![](fig/instance_segmentation_clear_border.png){
+alt="The instance segmentation with any nuclei crossing the image boundary
+removed"}
+We now have an image with 11 clearly labelled nuclei. Let's check the
+nuclei count as we did above.
+
+```python
+# First count the nuclei
+number_of_nuclei = instance_seg.max()
+print(f"There are {number_of_nuclei} individual nuclei")
+```
+```output
+There are 19 individual nuclei
+```
+Why are there still 19 nuclei? When we ran `clear_borders` the pixels
+corresponding
+to border nuclei were set to zero, however the total number of labels
+in the image was not changed, so whilst there are 19 labels in the
+image some of them have no corresponding pixels. The easiest way to
+correct this is to re label the image.
+```python
+instance_seg = label(instance_seg)
+number_of_nuclei = instance_seg.max()
+print(f"There are {number_of_nuclei} individual nuclei")
+```
+```output
+There are 11 individual nuclei
+```
+
+You now have a correct instance segmentation. You could return to
+using the napari-skimage-regionprops plugin to calculate the sizes
+of each nucleus and export the results to a speadsheet or your preferred
+analysis software using the `save as csv` function. However you've probably
+picked up enough Python during this course to complete the analysis you need
+with just the Napari console. Let's give it a try. The following commands
+should work with copy and paste, so don't worry too much if you don't think
+you've quite mastered Python for loops yet.
+
+Earlier in the lesson we used a Python [for loop](
+https://swcarpentry.github.io/python-novice-inflammation/05-loop.html)
+to try out three different erosion radii. Now let's use a for loop to
+count the number of pixels in each of the 11 nuclei. For this we'll
+take advantage of Python's `range` function which returns a list of numbers.
+
+```python
+
+# Create a list label values for each label in the instance segmetation
+# We need to add 1 the number of nuclei as python's range function
+# includes the lower limet (1) but not the upper limit.
+labels = range(1, number_of_nuclei + 1)
+
+# Print it to the console to check it's right
+for label in labels:
+  print (label)
+
+```
+```output
+1
+2
+3
+4
+5
+6
+7
+8
+9
+10
+11
+```
+
+Earlier in the lesson we used Numpy's `count_nonzero` function to find the
+size of
+a single nucleus. Let's put it inside the for loop to find the size of
+every nucleus. We'll use a Python list to store the results.
+
+```python
+# Create an empty list
+nucleus_pixels = []
+# Go through each nucleus,
+for label in labels:
+  # Append the number of pixels to the list
+  nucleus_pixels.append(np.count_nonzero(instance_seg == label))
+```
```

Comment:
Worth printing out the list at the end of the loop and explaining what it contains?

Reply:
Added with 3d3209a